### PR TITLE
Courier delivery details screen

### LINKED ISSUE #9: The screen for selection of home address
A user should have an oppotunity to fill details about their home address if they chose this delivery method.

- Date and time of delivery - required:
    - A user can pick up a date at least 3 days away from the today
    - A user should be able to pick up time of delivery. 
- Home address details:
    - City - required
    - Street - required
    - House - required
    - Entrance 
      - optional
      - should accept only positive integer
    - Flat:
      - optional
      - should accept only positive integer
    - Has elevator - required

### PR DESCRIPTION
**Pull request by coding-vibe**

### CODE REVIEW

File: src/utils/validation.ts [hunk 1]
```diff
@@ -36,6 +36,17 @@ export const validateIsFutureDate = (value: string) => {
   return isFuture(parsedDate) ? undefined : 'Card expired';
 };
 
+export const isPositiveInteger = (entity: string) => (value: string) => {
```
Comment (singvarr) [inline src/utils/validation.ts]: Shouldn't be a string. Find API that formats value before the moment when it's written to the form state.

File: src/utils/validation.ts [hunk 2]
```diff
@@ -36,6 +36,17 @@ export const validateIsFutureDate = (value: string) => {
...
+  const positiveInteger = /^(?:[0-9]+\/[0-9]+|[0-9]+|[a-zA-Z]+|\/)+$/;
```
Comment (singvarr) [inline src/utils/validation.ts]: Also make me simpler: use `isInteger` from `lodash` and check that's bigger than 0

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  return (
+    <Form<CourierDeliveryDetailsType>
+      onSubmit={(values) => {
+        onSaveFormValues(FormScreens.COURIER_DELIVERY_DETAILS, values);
+      }}
+      render={({ handleSubmit }) => (
+        <form onSubmit={handleSubmit}>
+          <Box sx={{ mb: 2 }}>
+            <span>Choose date and time for courier delivery</span>
```
Comment (singvarr) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx]: Unite date and time in `fieldset`

File: src/utils/validation.ts [hunk 4]
```diff
@@ -36,6 +36,17 @@ export const validateIsFutureDate = (value: string) => {
...
+
+  return positiveInteger.test(value) ? undefined : `Invalid ${entity} number`;
+};
+
+export const validateMinDate = (value: Date) =>
+  differenceInDays(value, new Date()) >= 2
```
Comment (singvarr) [inline src/utils/validation.ts]: Don't hardcode 2 in the validator and error message

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            name='date'
+            sx={{ mb: 2 }}
+          />
+          <TimePicker
+            fieldProps={{ validate: validateIsRequired }}
+            label='Time'
+            name='time'
+            sx={{ mb: 2 }}
+          />
+          <Box sx={{ mb: 2 }}>
```
Comment (singvarr) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx]: Also wrap remaining fields in `fieldset`

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+          />
+          <TextField
+            fieldProps={{ validate: isPositiveInteger('intercom') }}
+            label='Intercom'
+            name='intercom'
+            placeholder='Enter intercom number'
+            sx={{ mb: 2 }}
+          />
+          <Box sx={{ mb: 2 }}>
+            <Select
```
Comment (singvarr) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx]: I should be checkbox

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+export const MIN_DATE = addDays(new Date(), 3);
+
+type CourierDeliveryDetailsType =
+  InitialFormValuesType[FormScreens.COURIER_DELIVERY_DETAILS];
+
+export default function DeliveryDetailsScreen() {
+  const { formValues, onSaveFormValues } = useContext(WizardFormContext);
+  console.log(formValues);
+
```
Comment (singvarr) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx]: Don't forget to remove `console.log`

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  InitialFormValuesType,
+} from 'contexts/WizardFormContext';
+import {
+  composeValidators,
+  isPositiveInteger,
+  validateMinDate,
+  validateIsRequired,
+} from 'utils/validation';
+
+export const MIN_DATE = addDays(new Date(), 3);
```
Comment (singvarr) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx]: Rename `MIN_DATE` to `MIN_DELIVERY_DATE`

Also move 3 to const

File: src/utils/validation.ts (hunk 1 above)
Comment (singvarr) [reply src/utils/validation.ts]: Also rename me to `validateIsPositiveInteger`

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            defaultValue={MIN_DATE}
+            fieldProps={{
+              validate: composeValidators(validateIsRequired, validateMinDate),
+            }}
+            label='Date'
+            minDate={MIN_DATE}
+            name='date'
+            sx={{ mb: 2 }}
+          />
+          <TimePicker
```
Comment (singvarr) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx]: Extract time on submit and save value to the same format as it's displayed for the user.

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+      }}
+      render={({ handleSubmit }) => (
+        <form onSubmit={handleSubmit}>
+          <Box sx={{ mb: 2 }}>
+            <span>Choose date and time for courier delivery</span>
+          </Box>
+          <DatePicker
+            defaultValue={MIN_DATE}
+            fieldProps={{
+              validate: composeValidators(validateIsRequired, validateMinDate),
```
Comment (singvarr) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx]: Add `parse` handler that will convert `string` to `Date` and eliminate this error

File: src/utils/validation.ts [hunk 11]
```diff
@@ -1,5 +1,6 @@
-import { isFuture } from 'date-fns';
+import { differenceInDays, isFuture } from 'date-fns';
 import { FieldValidator } from 'final-form';
+import { isInteger } from 'lodash';
```
Comment (coding-vibe) [inline src/utils/validation.ts]: Import it better

File: package.json [hunk 12]
```diff
@@ -16,9 +16,12 @@
     "@emotion/react": "^11.11.1",
     "@emotion/styled": "^11.11.0",
     "@mui/material": "^5.14.12",
+    "@mui/x-date-pickers": "^6.17.0",
+    "@types/lodash": "^4.14.200",
```
Comment (coding-vibe) [inline package.json]: Move me to `devDependencies`

File: src/main.tsx [hunk 13]
```diff
@@ -1,14 +1,19 @@
...
 import Routes from 'components/Routes';
+
 import WizardFormProvider from 'components/WizardFormProvider';
 
 ReactDOM.createRoot(document.getElementById('root')!).render(
   <StrictMode>
     <BrowserRouter>
       <WizardFormProvider>
-        <Routes />
+        <LocalizationProvider dateAdapter={AdapterDateFns}>
```
Comment (coding-vibe) [inline src/main.tsx]: Move me under `BrowserRouter` immediately

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+type CourierDeliveryDetailsType =
+  InitialFormValuesType[FormScreens.COURIER_DELIVERY_DETAILS];
+
+export default function DeliveryDetailsScreen() {
+  const { onSaveFormValues } = useContext(WizardFormContext);
+
+  return (
+    <Form<CourierDeliveryDetailsType>
+      onSubmit={(values) => {
```
Comment (coding-vibe) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx:34]:
```suggestion
    <Form<Omit<CourierDeliveryDetailsType, 'time'> & { time: Date };>
```

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            />
+            <TextField
+              fieldProps={{
+                format: formatHandler,
+                parse: parseHandler,
+              }}
+              label='Intercom'
+              name='intercom'
+              placeholder='Enter intercom number'
+              sx={{ mb: 2 }}
```
Comment (coding-vibe) [inline src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx:115]: Should be in `fieldset`

File: src/utils/validation.ts [hunk 16]
```diff
@@ -36,6 +37,15 @@ export const validateIsFutureDate = (value: string) => {
   return isFuture(parsedDate) ? undefined : 'Card expired';
 };
 
+export const validateMinDate =
+  (count: number, message: string) => (value: Date) =>
```
Comment (coding-vibe) [inline src/utils/validation.ts]: Rename `count` to `deltaFromNowInDays`

File: src/contexts/WizardFormContext.tsx [hunk 17]
```diff
@@ -7,6 +7,16 @@ export const initialFormValues = {
     cvvCode: null,
     expirationDate: '',
   },
+  [FormScreens.COURIER_DELIVERY_DETAILS]: {
+    date: Date,
```
Comment (singvarr) [inline src/contexts/WizardFormContext.tsx]: Maybe new Date() or omit the field at all?

File: src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx (hunk 14 above)
Comment (singvarr) [reply src/pages/CourierDeliveryDetailsScreen/CourierDeliveryDetailsScreen.tsx:34]: Extract to a type above the component

File: src/main.tsx [hunk 18]
```diff
@@ -1,15 +1,23 @@
...
 
 ReactDOM.createRoot(document.getElementById('root')!).render(
   <StrictMode>
     <BrowserRouter>
-      <WizardFormProvider>
-        <Routes />
-      </WizardFormProvider>
+      <LocalizationProvider dateAdapter={AdapterDateFns}>
+        <WizardFormProvider>
+          <CssBaseline />
```
Comment (singvarr) [inline src/main.tsx]: Should be above the wizard form

**Review by singvarr — APPROVED** [review]

**Pull request merged**